### PR TITLE
Use requests for downloading, rich for progress bars & concurrently download packages

### PR DESCRIPTION
**Pull request by joebonrichie**
## Summary

Switches fetcher to use requests for downloading and rich for the progress bars
Adds a `fetch_multi()` method to download multiple packages concurrently
Changes various eopkg operations to use `fetch_multi()` for concurrent downloads

## TODO

- Takeback needs testing against obsolete packages (how?)

## Test Plan

Create a new rootfs and install system.base, system.devel as well as random packages to it
Fetch single package
Fetch multiple packages
Nuke cache then takeback to a state
Install package from a local repo
Fetch packages from local repo and remote repo
Do the python 3.14 rebuilds against this branch

### CODE REVIEW

File: pisi/fetcher.py [hunk 1]
```diff
@@ -1,319 +1,210 @@
-# SPDX-FileCopyrightText: 2005-2011 TUBITAK/UEKAE, 2013-2017 Ikey Doherty, Solus Project
+# SPDX-FileCopyrightText: 2026 Solus Project
 # SPDX-License-Identifier: GPL-2.0-or-later
```
Comment (EbonJaeger) [inline pisi/fetcher.py:2]: We should fix this. That was a result of me starting this in a new file, and then removing the old `fetcher.py`  and coping the new one to the same name.

File: pisi/operations/history.py [hunk 2]
```diff
@@ -49,84 +48,69 @@ def __listactions(actions):
...
     except PackageNotFound:
         errors.append(package)
         ctx.ui.info(pisi.util.colorize(_("%s could not be found") % (package), "red"))
         return False
 
-    dest = ctx.config.cached_packages_dir()
-    filepath = os.path.join(dest, uri.filename())
-    if not os.path.exists(filepath):
-        failed = False
+    # TODO(Evan): Validate
```
Comment (EbonJaeger) [inline pisi/operations/history.py]: I don't remember why I put this here. Nothing in the current diff jumps out at me.

File: pisi/api.py [hunk 3]
```diff
@@ -428,23 +427,30 @@ def fetch(packages=[], path=os.path.curdir, repo=None):
...
+        ) and resource.expected_hash == pisi.util.sha1_file(resource.local_path):
             ctx.ui.warning(_(f"{resource.uri.filename()} package already fetched"))
             continue
 
-        fetcher.fetch_url(resource.uri, path, ctx.ui.Progress)
+        fetch_items.append(resource)
+
+    # TODO(Evan): Error handling
+    if fetch_items:
+        fetcher.fetch_multi(fetch_items)
```
Comment (EbonJaeger) [inline pisi/api.py]: How should we handle this TODO? The old code didn't do any sort of error handling, so is the caller just expected to do it?

File: pisi/sourcearchive.py [hunk 4]
```diff
@@ -68,11 +70,13 @@ def fetch(self, interactive=True):
                 % (ctx.config.archives_dir(), self.url.filename())
             )
 
+    # TODO(Evan): WTF
```
Comment (EbonJaeger) [inline pisi/sourcearchive.py]: This can probably be removed.

File: pisi/api.py (hunk 3 above)
Comment (joebonrichie) [reply pisi/api.py]: It's probably fine here specifically to allow individual downloads to 404 here since we're only fetching.

File: pisi/operations/history.py (hunk 2 above)
Comment (joebonrichie) [reply pisi/operations/history.py]: Alright, I've restored the previous behaviour

Comment (EbonJaeger) [reply pisi/operations/history.py]: I only meant my todo comment; the code you had was fine

File: pisi/sourcearchive.py [hunk 5]
```diff
@@ -82,14 +85,12 @@ def fetch_from_mirror(self):
             try:
                 url = os.path.join(mirror, archive)
                 ctx.ui.warning(_("Fetching source from mirror: %s") % url)
-                pisi.fetcher.fetch_url(url, ctx.config.archives_dir(), self.progress)
+                fetcher.fetch(url, ctx.config.archives_dir())
                 return
-            except pisi.fetcher.FetchError:
+            except HTTPError | IOError | ValueError:
```
Comment (EbonJaeger) [inline pisi/sourcearchive.py:91]: I just noticed that this syntax is different from other places. Should it be like this instead?

```python
except (HTTPError, IOError, ValueError):
```

File: pisi/operations/build.py [hunk 6]
```diff
@@ -530,11 +526,13 @@ def fetch_component(self):
...
-        progress = ctx.ui.Progress
         if pisi.uri.URI(url).is_remote_file():
+            fetcher = Fetcher()
+
+            # TODO(Evan): wat
             try:
-                pisi.fetcher.fetch_url(url, self.pkg_work_dir(), progress)
-            except pisi.fetcher.FetchError:
+                fetcher.fetch(url, self.pkg_work_dir())
+            except HTTPError | IOError | ValueError:
```
Comment (EbonJaeger) [inline pisi/operations/build.py:535]: Same thing here.

File: pisi/sourcearchive.py (hunk 5 above)
Comment (joebonrichie) [reply pisi/sourcearchive.py:91]: PEP 604 (py 3.10), but we should probably be consistent yeah

**Comment by sheepman4267** [comment]
I *love* this. It's so much better. The concurrency is so clean and easy to understand, and the resulting terminal output is much more compact and reliable. Thank you SO MUCH for implementing this. 

I notice one feature that's no longer present, though- the old UI would show which package it was on, providing some semblance of an overall progress indicator. Could this be trivially added back, or should that be another future PR? I'm happy either way, and I want this to get in without a lot of delays, but I think it's an important feature. I'm thinking of an additional progress bar at the bottom of the stack of progress bars (probably a different color?). Tracking how many packages are downloaded would probably be easiest, but for extra credit a by-total-download-size progress bar would also be quite cool.

**Comment by sheepman4267** [comment]
(Or just add the `(3/162)` thing to the "downloaded" message per package.

**Comment by joebonrichie** [comment]
Yes, i've been thinking it'll be nice to implement a global progress bar at the bottom as well, i'll need to brush up my terminal knowledge as well as dive into rich's documentation to figure out how to do that though. I'll have a think if there's any easy way to add back a global counter somewhere that's intuitive in the meanwhile

**Comment by EbonJaeger** [comment]
I wonder if this PR is getting a bit too big (in both size and scope), and if we shouldn't instead iterate on it in subsequent PRs.

**Comment by sheepman4267** [comment]
> I wonder if this PR is getting a bit too big (in both size and scope), and if we shouldn't instead iterate on it in subsequent PRs.

This is probably a very good idea.

**Comment by sheepman4267** [comment]
However, I now have a critique that may actually want to be addressed here (though it's not a blocker; I know it's probably going to be hard or impossible to solve). I put my laptop to sleep earlier while eopkg was downloading updates, and when I came back, I had an absolutely incredible number of errors taking up many screens of text in my terminal: https://gist.github.com/sheepman4267/ce2fdba4c8a2368d321f67dddda2c12a

I can see why this happened, and it's not a huge deal, but I think there is probably a nicer way to present messages about connection loss to users.

**Comment by joebonrichie** [comment]
Yeah two bugs there, stale terminal escape sequence getting printed to the console somehow `[0m`, that's gonna be annoying af to track down.

I guess now we have 5 * 8 retries by default, we could consider making retries global instead of per-worker

**Comment by sheepman4267** [comment]
I found another (related, I think) bug: https://gist.github.com/sheepman4267/48e9bd82fbd12105935d8c8dddd48916

In this case, I switched WiFi networks during the fetch phase. It looks like some of the in-progress fetches failed (duh) and didn't get retried when the network came back. Other packages were downloaded after the network came back, though. 

Does each fetcher process manage its own set of retries? Is it possible that when one fetcher runs out of retries, *only that fetcher* dies while the others keep downloading the rest of the packages? I haven't looked closely at the code and don't have the mental space right now, so I'm sorry but I will just speculate :)

Also- just watching the terminal output, I'm not sure that the retry logic for waiting between retries is working. It feels like the retries all happen immediately, which is fine in the case of a flaky connection where a package downloads a bit, fails, then downloads a bit... but it's not that helpful when the connection goes away entirely. If it would wait in exponentially increasing numbers of seconds, I think that would be better.

If that last part is true, though, it should probably be a separate issue. Just something related that I noticed in testing.

Good news, though: It works! Due to my own network messing-around, I did have to restart it twice, but it resumes and continues properly, it all works, and it's fast. *damn*, it's fast. The reworked logic around displaying/detecting cached packages is much appreciated and way, way, way faster.

**Comment by joebonrichie** [comment]
> Does each fetcher process manage its own set of retries? Is it possible that when one fetcher runs out of retries, only that fetcher dies while the others keep downloading the rest of the packages? I haven't looked closely at the code and don't have the mental space right now, so I'm sorry but I will just speculate :)

Hmm, yeah i think i changed that recently in commits [c8f774c](https://github.com/getsolus/eopkg/pull/315/commits/c8f774c975ca6d0e12f11021f1fd6f6667c43c1a) and [4c2b1ba](https://github.com/getsolus/eopkg/pull/315/commits/4c2b1ba40ba35ec2d43694fdc9cc273695181387) to explicitly allow historical package downloads to fail so it would reach the existing failure case (e.g. packages trimmed from repo) 

> In this case, I switched WiFi networks during the fetch phase. It looks like some of the in-progress fetches failed (duh) and didn't get retried when the network came back. Other packages were downloaded after the network came back, though.

brother, really testing the edge-cases huh, I wonder if the connection pooling fails if you switching networking mid-flight.

Okay, tbh most of these are nits which definitely can be improved, in terms of priority of things that can be improved:
- Fix the stale terminal escape sequence (`[0m`) that gets printed in case of a error (timeout, etc. why the fuck does this only happen in the `ThreadPoolExecutor` context but not elsewhere????)
- Consider tweaking retry logic, previously we had 5 retries per download but now we have 8 download workers by default we should consider either using a global retry counter or possibly lowering the default retries to 3.
- Add a global progress bar, this is definitely possible with rich but need to dive into the documentation (live display?)

**Comment by joebonrichie** [comment]
> Also- just watching the terminal output, I'm not sure that the retry logic for waiting between retries is working. It feels like the retries all happen immediately, which is fine in the case of a flaky connection where a package downloads a bit, fails, then downloads a bit... but it's not that helpful when the connection goes away entirely. If it would wait in exponentially increasing numbers of seconds, I think that would be better.

Yeah it's just 15 seconds by default, haven't really changed the logic from the previous sequential implementation, but a back-off could be implemented

**Comment by sheepman4267** [comment]
I mean, I didn't test the edge cases on purpose, I just wanted to switch from bad conference WiFi to my phone hotspot to speed the thing up. In fact, I didn't mean to do either of these- they just happened while I was trying to run updates :)

Anyway, I think that your second and third points of possible improvement should be made into issues to work on in the future. This PR is already huge, and those (especially point 3) can probably be considered out of scope.

**Comment by joebonrichie** [comment]
Fixed the stale terminal escape sequence ([0m) that gets printed in case of a error

**Comment by EbonJaeger** [comment]
At the risk of opening this up to Bikeshedding Hell, I wonder if yellow is the best color to use for the `Downloaded thinger.bob` output. To me, it looks too much like a warning, when really, it's an informational line.

**Comment by joebonrichie** [comment]
> At the risk of opening this up to Bikeshedding Hell, I wonder if yellow is the best color to use for the `Downloaded thinger.bob` output. To me, it looks too much like a warning, when really, it's an informational line.

Hmm i just followed the pre-existing eopkg colour scheme, I could change it to green if peeps think that's better

**Review by EbonJaeger — DISMISSED** [review]
LGTM, thanks!
